[#134] FEATURE: import and export filters

jacksonjacobs1 opened the pull request:
https://jacksonjjacobs.com/openproject/work_packages/134

jacksonjacobs1 committed:
[#134] FEATURE: import and export filters

https://jacksonjjacobs.com/openproject/work_packages/134

jacksonjacobs1 reviewed (commented):
## Description 
1. Traithorizon does not have any method for setting filters other than by dragging along parcoords axes. This is not particularly precise and does not allow for the case where a user has an existing set of thresholds defined for their use case.
2. Traithorizon does not have any method for exporting the currently-set filter thresholds, impeding the reproducibility of filters.

## Implementation
This feature will center around supporting a json-based format for representing filters, based on the dict format documented in https://github.com/BigFatDog/parcoords-es

1D Brush
```node
  .brushMode("1D-axes")
  .brushExtents({"2": [3,4]});
```

Multi Brush
```node
  .brushMode("1D-axes-multi")
  .brushExtents({"2": [[3,4], [6,8], [10, 14]]});

```

### Import implementation
Filters will be imported via a JSON file can be dropped directly onto the parcoords plot.

There will also be an export filters button available on the parcoords plot. Clicking this button will result in a JSON file containing the filters being prepared for download to the user's local machine.

choosehappy commented:
is file based the right way to go - or would it be better to have "import from clipboard" and "export to clipboard"?

jacksonjacobs1 commented:
I guess it depends on how the user would naturally want to store the application state, and I could see this going both ways. Maybe if they want to store a bunch of configurations it would be more convenient to use their file system rather than pasting a bunch of configurations in their notes?

Do you have any reasons in mind why clipboard would be preferred? My sense is that both of these solutions have similar implementation time.

choosehappy commented:
i feel like i've seen clipboard based in some other tools, but they don't come to mind at the moment. regardless, in thinking more about it, i think you're right, and if the implementation time is similar, files are likely a much smarter way to go!

jacksonjacobs1 committed:
Add Dropzone for filter JSON file upload in index.html

- Included Dropzone CSS and JS files for file upload functionality.
- Added a Dropzone area in the card body for users to drop or click to upload filter JSON files.

jacksonjacobs1 committed:
Update Python Debugger configuration to include arguments for module execution

jacksonjacobs1 committed:
Finish import and export filter thresholds functionality

jacksonjacobs1 commented:
Done:

the updated UI looks as follows with minimal additional formatting:
<img width="4690" height="2506" alt="image" src="https://github.com/user-attachments/assets/1beb8ce3-e7f3-43e5-a68e-77dbc5ab62a9" />

After adding brushed filters, click on the "Export Filter Extents" button to export a json file:
<img width="4698" height="1688" alt="image" src="https://github.com/user-attachments/assets/6af24bb8-92ae-452a-b438-6b966785664f" />

```json
{
  "TE_THICK_MIN": [
    0.43035547300000054,
    3.1208903418664473
  ],
  "LUMEN_THICK_MAX": [
    1.879044876030676,
    6.2307925082019295
  ],
  "LUMEN_THICK_AVE": [
    1.049596475281823,
    2.2814853265295296
  ],
  "LUMEN_THICK_STD": [
    0.6264710734918291,
    0.883690952904673
  ]
}
```

Note that by changing the `BRUSH_MODE` global variable to `BRUSH_MODES.ONE_D_AXES_MULTI` results in the ability to add multiple brush filters on a single axis. This is not the default behavior, but it is now supported by traithorizon and results in a slightly different looking json file (note extra nesting of arrays), as described in the [parcoords documentation](https://github.com/BigFatDog/parcoords-es?tab=readme-ov-file).
<img width="4678" height="2358" alt="image" src="https://github.com/user-attachments/assets/e166b017-d0f1-40a1-a57a-f46459011d78" />


```json
{
  "TUBULE_THICK_STD": [
    [
      5.556425334751463,
      7.083021384962237
    ],
    [
      2.6813361068545074,
      4.3097052270793315
    ]
  ],
  "LUMEN_THICK_MAX": [
    [
      22.139579162744806,
      27.429886128992976
    ],
    [
      11.757239085882977,
      16.1793468963982
    ],
    [
      3.489820135789298,
      8.360547579255343
    ]
  ]
}
```

After obtaining a filter extents json file, you can drag it directly into the UI, resulting in the filters being set programatically (note this image looks the same as the last one, because the filters were correctly reproduced):
<img width="4698" height="2368" alt="image" src="https://github.com/user-attachments/assets/526184b8-8688-44f2-bc11-3ea7fb4a7c00" />

Please let me know if any any formatting / functionality changes need to be made.

jacksonjacobs1 committed:
Refactor brush mode handling and enhance multi-axes support in parcoords filter extents

jacksonjacobs1 commented:
@choosehappy thoughts on combining this feature with #13 , so that we have a single export button resulting in the following json file format:

```json
{
    "extents": {
      "TUBULE_THICK_AVE": [
        17.496402000911264,
        30.209967107267232
      ]
    },

    "filtered_row_ids": [0, 34, 53]
}
```

The "import" use case would then use this file but only look at the "extents" data. The filtered row ids would be irrelevant in that context.

choosehappy commented:
yup, i like it. would probably make the import button clear however "Import Settings" versus "Export Settings and IDs", so its clarified to the user what will be happening

relatedly, - whats the motivation for calling it "Extents" and not "Settings" - regardless the button text should line up with the export value (e.g., both settings, or both extents, but not a extents/settings)

choosehappy commented on traithorizon/static/js/slickgrid.js:
```diff
@@ -25,12 +28,39 @@ $(document).ready(function () {
+							var extents = JSON.parse(e.target.result);
+							setParcoordsFilterExtents(extents);
+						} catch (err) {
+							alert("Invalid JSON file: " + err);
```
agree with this feedback

choosehappy commented on traithorizon/static/js/slickgrid.js:
```diff
@@ -25,12 +33,39 @@ $(document).ready(function () {
+			acceptedFiles: ".json",
+			maxFiles: 1,
+			addRemoveLinks: true,
+			dictDefaultMessage: "Drop filter JSON file here or click to upload",
```
"filter" may not be the best word, shoudl align with the comment i made in the issue, should all be extents or settings

choosehappy commented on traithorizon/static/js/slickgrid.js:
```diff
@@ -569,26 +645,31 @@ function tooltip(selectionGroup, tooltipDiv) {
+		const deviation = numbers.length > 1 ? d3.deviation(numbers).toFixed(fixedPrecision) : undefined;
+
+		// Calculate mean only if there are at least 1 numbers
+		const mean = numbers.length > 0 ? d3.mean(numbers).toFixed(fixedPrecision) : undefined;
```
this seems like good feedback as well?

jacksonjacobs1 committed:
Implement general feedback

jacksonjacobs1 committed:
- Refactor settings import functionality and update Dropzone integration in SlickGrid.
- Export filtered row ids within settings.json file
- Rename "extents" to "settings"

jacksonjacobs1 committed:
Remove color styling from filter toggle button in index.html

jacksonjacobs1 commented:
Should be done now! The UI now looks like this:
<img width="4162" height="2420" alt="image" src="https://github.com/user-attachments/assets/15bae264-2e6d-4218-9906-3cf09b8f8091" />

The output format of the JSON file looks like this:
```json
{
   "filterSettings": {
      "TBM_THICK_AVE": [
         1.7356824961876072,
         3.6997556957120628
      ],
      "TUBULE_THICK_STD": [
         1.8448265703242175,
         6.007497240389435
      ]
   },
   "filteredRowIds": [
      7,
      12,
      15,
      16
   ]
}
```

Note that this settings file leaves the opportunity to store additional configurable settings in the future. It's just a bit conceptually confusing that filteredRowIds are contained in the settings file.

jacksonjacobs1 committed:
Add jsonify import to routes.py for JSON response handling

jacksonjacobs1 committed:
update UI text

jacksonjacobs1 committed:
Adjust layout

jacksonjacobs1 commented:
Updated layout:
<img width="4704" height="2512" alt="image" src="https://github.com/user-attachments/assets/107bf61c-e45e-467e-b203-5e4a2d6a6072" />

choosehappy commented:
hot.
